fog/ledger/server/tests/router_connection.rs

```diff
@@ -863,6 +863,281 @@ fn fog_ledger_untrusted_tx_out_api_test(logger: Logger) {
     sleep(Duration::from_millis(1000));
 }
 
+// Test that a fog ledger connection is able to check key images by hitting
+// a fog ledger router using the unary API
+#[test_with_logger]
+fn fog_router_unary_key_image_test(logger: Logger) {
+    let mut rng = RngType::from_seed([0u8; 32]);
+
+    for block_version in BlockVersion::iterator() {
+        let alice = AccountKey::random_with_fog(&mut rng);
+
+        let recipients = vec![alice.default_subaddress()];
+
+        let keys: Vec<KeyImage> = (0..20).map(|x| KeyImage::from(x as u64)).collect();
+
+        // Make LedgerDB
+        let ledger_dir = TempDir::new("fog-ledger").expect("Could not get test_ledger tempdir");
+        let db_full_path = ledger_dir.path();
+        let mut ledger = recreate_ledger_db(db_full_path);
+
+        // Make WatcherDB
+        let (mut watcher, watcher_dir) = setup_watcher_db(logger.clone());
+
+        // Populate ledger with some data
+        // Origin block cannot have key images
+        add_block_to_ledger(
+            block_version,
+            &mut ledger,
+            &recipients,
+            &[],
+            &mut rng,
+            &mut watcher,
+        );
+        add_block_to_ledger(
+            block_version,
+            &mut ledger,
+            &recipients,
+            &keys[0..2],
+            &mut rng,
+            &mut watcher,
+        );
+        add_block_to_ledger(
+            block_version,
+            &mut ledger,
+            &recipients,
+            &keys[3..6],
+            &mut rng,
+            &mut watcher,
+        );
+        let num_blocks = add_block_to_ledger(
+            block_version,
+            &mut ledger,
+            &recipients,
+            &keys[6..9],
+            &mut rng,
+            &mut watcher,
+        );
+
+        // Populate watcher with Signature and Timestamp for block 1
+        let url1 = Url::parse(TEST_URL).unwrap();
+        let block1 = ledger.get_block(1).unwrap();
+        let signing_key_a = Ed25519Pair::from_random(&mut rng);
+        let filename = String::from("00/00");
+        let mut signed_block_a1 =
+            BlockSignature::from_block_and_keypair(&block1, &signing_key_a).unwrap();
+        signed_block_a1.set_signed_at(1593798844);
+        watcher
+            .add_block_signature(&url1, 1, signed_block_a1, filename.clone())
+            .unwrap();
+
+        // Update last synced to block 2, to indicate that this URL did not participate
+        // in consensus for block 2.
+        watcher.update_last_synced(&url1, 2).unwrap();
+
+        {
+            // Make Key Image Store
+            let store_uri = KeyImageStoreUri::from_str(&format!(
```

Comment:
It appears that this test is only involving one `KeyImageStore` acting as a shard. We should test this scenario but we also need to test the scenario in which there are multiple shards. Are there plans to do so in a follow up PR?

Reply:
Yes - I suspect this would fit better in a separate PR and is out-of-scope for this one, but it's very necessary to test that behavior because only having one store is so different from what the router would actually be doing in production. Is it safe to say this is out of scope for this PR, you think? Should I make an issue for this?

Reply:
Per discussion this looks safe to bring forward into a separate PR and move out of this one.